### ISSUE TITLE
Editing subtly corrupts a message file

### ISSUE THREAD

**Issue by humdingerb** [andimachovec/Kottan#16]
The "MediaFile" of the Sounds preferences saves it's settings in `~/config/settings/Media/MediaFiles` as BMessage. I recently wanted to edit the path to sound files with Kottan, but the result, for some reason, didn't work with the Sounds preferences.

Kottan opens the file OK and shows the edited path correctly, but something seems to be amiss. In Terminal, "message" shows the file correctly, but ends with:

`failed to unflatten message: Data is not a message`

Attached it the working original and "MediaFiles_edited" which was edited (".wav" to ".flac").
[MediaFiles.txt](https://github.com/user-attachments/files/18198420/MediaFiles.txt)
[MediaFiles_edited.txt](https://github.com/user-attachments/files/18198421/MediaFiles_edited.txt)

(The ".txt" suffix is needed to upload on github...)

**Comment by andimachovec** [comment]
Thanks for reporting. I can reproduce the problem here, but not consistently. In about 1/3 of my tries it works OK (at least the message command doesn't complain). 
The flattening seems to work (at least it returns B_OK) , but sometimes the message data in the main BMessage object appears to be corrupted. 
Can you reproduce the error consistently on your system? 
I'll publish a branch with debug output tomorrow so we can do more in-depth testing.

**Comment by humdingerb** [comment]
A bit more digging...
It appears that keeping the string the same length doesn't corrupt the file, e.g.
```
/boot/system/data/sounds/burgs-sparkly-tones/Beep.ogg ->
/boot/system/pata/sounds/burgs-sparkly-tones/Beep.ogg
```

Changing the legth by removing "data" or changing it to "patata", does produce the corruption.
Also, once corrupted, reverting the change (e.g. "patata"-> "data") 'un-corrupts' the file.

(All in regard to the "message" command output, I didn't test the file with the actual Sounds prefs.)

**Comment by humdingerb** [comment]
Hmm... strike the above. It appeared consistently as described, but now it doesn't anymore.
Heisenbugger. :\

**Comment by andimachovec** [comment]
> Heisenbugger. :\

Ah ... those are the most fun to debug ;-)
I've been a bit extra busy with preparations for Christmas in the last few days, so not much progress with debugging. But I have a suspicion it could have something to do with my handling of the BFile object used to unflatten and flatten the messages. 
We'll know in a few day when I have time to look at it. 
Merry Christmas and all the best, 
Andi (BlueSky)

**Comment by humdingerb** [comment]
Good hunting! And happy Xmas!

**Comment by andimachovec** [comment]
Alrighty, that took a bit longer than expected. Sorry for keeping you waiting. I still haven't completely isolated the bug, but here's what I know so far:
 I suspected the problem to be the fact that I kept the BFile object open (in B_READ_WRITE) mode, and "rewinded" the file with a Seek() call to an index of 0. (should be the beginning). Now I created a test branch where I only keep the entry_ref to the message file and construct (and delete) a new BFile object each time Kottan does file I/O. Unexpectedly, this does not change anything in relation to our problem ;-) 
I added some code to additionally flatten the BMessage object to a temporary file which results in a correct message file that even the message command doesn't complain about. This leads me to suspect that there is some problem when overwriting an existing message file. But I still have no clue why. I'll write a small test program to mimick the behaviour of Kottan's message handling, but shorten the turnaround time when testing. 
I'll keep you updated. 

Oh, and I'm not sure if it still counts in week 2, but Happy New Year!

**Comment by humdingerb** [comment]
So your bug hunt turned out to be a prolonged bug safari...

> Oh, and I'm not sure if it still counts in week 2, but Happy New Year!

I expect that 2025 has a lot for us in store that best wishes will be welcome all year round. Happy New Year, then. :)

**Comment by andimachovec** [comment]
> I expect that 2025 has a lot for us in store

You might be right on this one ;-)

> So your bug hunt turned out to be a prolonged bug safari...

Yeah, I started to expect it to become a full blown expedition ;-)
But, in a late night session last night (yes, I am slightly embarrassed I didn't have anything better to do on a saturday night) I managed to isolated the bug. It always appears when you change a string (or any data type of variable length) to a value shorter than the one before. The way it was implemented in Kottan, I would rewind the file descriptor to zero and start overwriting the data in the file. But, if the amount of new data is smaller, parts of the old and now useless data would remain at the end of the file and cause the problem. 
The solution is to open the BFile with the flag B_ERASE_FILE added to the open mask, so all data in the file gets deleted before writing the new data. Potentially a one line fix, but because I had used the same BFile object also for read operations I had to change a bit more code. 
You can test the new version by checking out the fix_flatten_bug branch. 
I did some rather quick testing yesterday and it looks good but it's always good to have an extra pair of eyes on it. I hope I didn't introduce any new bugs ;-)

**Comment by humdingerb** [comment]
Yep, seems to be fixed. I tried changing entries for a shorter and longer string, after saving the "message" command doesn't complain any more. Well done!

I found another issue while playing around. Report coming up...:)
